### ISSUE TITLE
0.40.0: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by christianjedroCDT** [grafana/agent#6557]
### What's wrong?

I´m using the new syntax (0.40) and i´m getting panic. I´m using `import.file` with the new "module call" like [this](https://grafana.com/docs/agent/latest/flow/concepts/modules/#:~:text=helpers.log_filter%20%22default%22%20%7B%0A%20%20//%20Configure%20the%20filter%20to%20forward%20filtered%20logs%20to%20loki.write%20below.%0A%20%20write_to%20%3D%20%5Bloki.write.default.receiver%5D%0A%7D) and i´m passing an argument inside this "call".

### Steps to reproduce

1. add config
2. start agent
3. get error

### System information

Linux

### Software version

Grafana Agent v0.40.0

### Configuration

main config
```
prometheus.remote_write "this" {
  endpoint {
    url = remote.kubernetes.configmap.config.data["PROMETHEUS_REMOTE_WRITE_URL"] + "/api/v1/write"

    basic_auth {
      username = remote.kubernetes.configmap.config.data["PROMETHEUS_REMOTE_USERNAME"]
      password = remote.kubernetes.secret.credentials.data["PROMETHEUS_REMOTE_PASSWORD"]
    }
... additional non related config ...
}
import.file "certmanager" {
  filename = "/etc/agent/certmanager.river"
  poll_frequency = "1m"
}
certmanager.config "this" {
  receiver = prometheus.remote_write.this.receiver
}
loki.write "this" {
  endpoint {
    name = "opsstack"
    url = remote.kubernetes.configmap.config.data["PROMETHEUS_REMOTE_WRITE_URL"] + "/loki/api/v1/push"

    basic_auth {
      username = remote.kubernetes.configmap.config.data["PROMETHEUS_REMOTE_USERNAME"]
      password = remote.kubernetes.secret.credentials.data["PROMETHEUS_REMOTE_PASSWORD"]
  }
... additional non related config ...
}
logging {
    level  = "warn"
    format = "json"
    write_to = [loki.write.this.receiver]
}
```

/etc/agent/certmanager.river

```
declare "config" {
    argument "receiver" {}

    discovery.kubernetes "cert_manager" {
      role = "endpoints"
      bearer_token_file = "/var/run/secrets/kubernetes.io/serviceaccount/token"

      selectors {
        role = "endpoints"
        label = "app.kubernetes.io/name=cert-manager,app.kubernetes.io/component=controller"
      }
    }

    discovery.relabel "cert_manager" {
      targets    = discovery.kubernetes.cert_manager.targets

      rule {
        source_labels = ["__meta_kubernetes_pod_container_port_name"]
        action        = "keep"
        regex         = "http-metrics"
      }

      rule {
        source_labels = ["__meta_kubernetes_endpoint_address_target_kind", "__meta_kubernetes_endpoint_address_target_name"]
        regex         = "Node;(.*)"
        target_label  = "node"
      }

      rule {
        source_labels = ["__meta_kubernetes_endpoint_address_target_kind", "__meta_kubernetes_endpoint_address_target_name"]
        regex         = "Pod;(.*)"
        target_label  = "pod"
      }

      rule {
        source_labels = ["__meta_kubernetes_namespace"]
        target_label  = "namespace"
      }

      rule {
        source_labels = ["__meta_kubernetes_service_name"]
        target_label  = "service"
      }

      rule {
        source_labels = ["__meta_kubernetes_pod_name"]
        target_label  = "pod"
      }

      rule {
        source_labels = ["__meta_kubernetes_pod_container_name"]
        target_label  = "container"
      }

      rule {
        source_labels = ["__meta_kubernetes_pod_phase"]
        regex         = "(Failed|Succeeded)"
        action        = "drop"
      }
    }

    prometheus.scrape "cert_manager" {
      targets    = discovery.relabel.cert_manager.output
      forward_to = [argument.receiver.value]
      job_name   = "cert-manager"
    }
}
```

### Logs

```text
{"ts":"2024-02-29T12:14:25.861873211Z","level":"info","boringcrypto enabled":false}
{"ts":"2024-02-29T12:14:25.862267086Z","level":"info","msg":"finished node evaluation","controller_id":"","trace_id":"021fec7ce794f129782a403b9a3ed8b8","node_id":"logging","duration":424166}
{"ts":"2024-02-29T12:14:25.86233542Z","level":"info","msg":"finished complete graph evaluation","controller_id":"","trace_id":"021fec7ce794f129782a403b9a3ed8b8","duration":29670875}
interrupt received
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xfddb1c]

goroutine 1 [running]:
github.com/grafana/agent/pkg/flow/logging.(*lokiWriter).Write(0x40000a8068?, {0x4002a5c400, 0xb6, 0x40010f7820?})
	/src/agent/pkg/flow/logging/logger.go:168 +0x6c
io.(*multiWriter).Write(0x0?, {0x4002a5c400, 0xb6, 0x400})
	/usr/local/go/src/io/multi.go:85 +0x74
github.com/grafana/agent/pkg/flow/logging.(*writerVar).Write(0xc1703b28736631bc?, {0x4002a5c400?, 0xcbcac20?, 0x40017c0930?})
	/src/agent/pkg/flow/logging/logger.go:218 +0xd4
log/slog.(*commonHandler).handle(_, {{0xc1703b28736631bc, 0x1038f9f4, 0xcbcac20}, {0x40017c0930, 0x22}, 0x0, 0x0, {{{0x40010f7810, 0xd}, ...}, ...}, ...})
	/usr/local/go/src/log/slog/handler.go:306 +0x550
log/slog.(*JSONHandler).Handle(_, {_, _}, {{0xc1703b28736631bc, 0x1038f9f4, 0xcbcac20}, {0x40017c0930, 0x22}, 0x0, 0x0, ...})
	/usr/local/go/src/log/slog/json_handler.go:91 +0x44
github.com/grafana/agent/pkg/flow/logging.(*handler).Handle(_, {_, _}, {{0xc1703b28736631bc, 0x1038f9f4, 0xcbcac20}, {0x40017c0930, 0x22}, 0x0, 0x0, ...})
	/src/agent/pkg/flow/logging/handler.go:49 +0x60
github.com/grafana/agent/internal/slogadapter.slogAdapter.Log({{0x82a92b0?, 0x4002ab7540?}}, {0x40029bc540, 0xa, 0xc0?})
	/src/agent/internal/slogadapter/gokit.go:90 +0x498
github.com/grafana/agent/pkg/flow/logging.(*Logger).Log(0x4002a98380, {0x40029bc540?, 0xa, 0xc})
	/src/agent/pkg/flow/logging/logger.go:158 +0x26c
github.com/go-kit/log.(*context).Log(0x40029eca00, {0x4002392d40?, 0x4, 0xc5a3f60?})
	/go/pkg/mod/github.com/dannykopping/go-kit-log@v0.2.2-0.20221002180827-5591c1641b6b/log.go:168 +0x31c
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).Apply.func1()
	/src/agent/pkg/flow/internal/controller/loader.go:175 +0x130
panic({0x605d140?, 0xc597780?})
	/usr/local/go/src/runtime/panic.go:914 +0x218
github.com/grafana/agent/pkg/flow/logging.(*lokiWriter).Write(0x40000a8068?, {0x4002a5c400, 0xbf, 0x40010f7740?})
	/src/agent/pkg/flow/logging/logger.go:168 +0x6c
io.(*multiWriter).Write(0x678e6?, {0x4002a5c400, 0xbf, 0x400})
	/usr/local/go/src/io/multi.go:85 +0x74
github.com/grafana/agent/pkg/flow/logging.(*writerVar).Write(0xc1703b28736526ce?, {0x4002a5c400?, 0xcbcac20?, 0x4002a9d710?})
	/src/agent/pkg/flow/logging/logger.go:218 +0xd4
log/slog.(*commonHandler).handle(_, {{0xc1703b28736526ce, 0x1037ef30, 0xcbcac20}, {0x4002a9d710, 0x18}, 0x0, 0x0, {{{0x40010f7710, 0xd}, ...}, ...}, ...})
	/usr/local/go/src/log/slog/handler.go:306 +0x550
log/slog.(*JSONHandler).Handle(_, {_, _}, {{0xc1703b28736526ce, 0x1037ef30, 0xcbcac20}, {0x4002a9d710, 0x18}, 0x0, 0x0, ...})
	/usr/local/go/src/log/slog/json_handler.go:91 +0x44
github.com/grafana/agent/pkg/flow/logging.(*handler).Handle(_, {_, _}, {{0xc1703b28736526ce, 0x1037ef30, 0xcbcac20}, {0x4002a9d710, 0x18}, 0x0, 0x0, ...})
	/src/agent/pkg/flow/logging/handler.go:49 +0x60
github.com/grafana/agent/internal/slogadapter.slogAdapter.Log({{0x82a92b0?, 0x4002ab7540?}}, {0x40029bc480, 0xc, 0xc0?})
	/src/agent/internal/slogadapter/gokit.go:90 +0x498
github.com/grafana/agent/pkg/flow/logging.(*Logger).Log(0x4002a98380, {0x40029bc480?, 0xc, 0xc})
	/src/agent/pkg/flow/logging/logger.go:158 +0x26c
github.com/go-kit/log.(*context).Log(0x40029ec960, {0x40029baa80?, 0x6, 0xc5a3f60?})
	/go/pkg/mod/github.com/dannykopping/go-kit-log@v0.2.2-0.20221002180827-5591c1641b6b/log.go:168 +0x31c
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).Apply.func2.1()
	/src/agent/pkg/flow/internal/controller/loader.go:188 +0x198
panic({0x605d140?, 0xc597780?})
	/usr/local/go/src/runtime/panic.go:914 +0x218
github.com/grafana/agent/pkg/flow/logging.(*lokiWriter).Write(0x40000a8068?, {0x4002a5c400, 0x54, 0x0?})
	/src/agent/pkg/flow/logging/logger.go:168 +0x6c
io.(*multiWriter).Write(0x0?, {0x4002a5c400, 0x54, 0x400})
	/usr/local/go/src/io/multi.go:85 +0x74
github.com/grafana/agent/pkg/flow/logging.(*writerVar).Write(0xc1703b28735f243b?, {0x4002a5c400?, 0xcbcac20?, 0x0?})
	/src/agent/pkg/flow/logging/logger.go:218 +0xd4
log/slog.(*commonHandler).handle(_, {{0xc1703b28735f243b, 0x1031ec73, 0xcbcac20}, {0x0, 0x0}, 0x0, 0x0, {{{0x4002a9d6e0, 0x14}, ...}, ...}, ...})
	/usr/local/go/src/log/slog/handler.go:306 +0x550
log/slog.(*JSONHandler).Handle(_, {_, _}, {{0xc1703b28735f243b, 0x1031ec73, 0xcbcac20}, {0x0, 0x0}, 0x0, 0x0, ...})
	/usr/local/go/src/log/slog/json_handler.go:91 +0x44
github.com/grafana/agent/pkg/flow/logging.(*handler).Handle(_, {_, _}, {{0xc1703b28735f243b, 0x1031ec73, 0xcbcac20}, {0x0, 0x0}, 0x0, 0x0, ...})
	/src/agent/pkg/flow/logging/handler.go:49 +0x60
github.com/grafana/agent/internal/slogadapter.slogAdapter.Log({{0x82a92b0?, 0x4002ab7540?}}, {0x4002a9a3c0, 0x4, 0x4001f0b1b0?})
	/src/agent/internal/slogadapter/gokit.go:90 +0x498
github.com/grafana/agent/pkg/flow/logging.(*Logger).Update(0x4002a98380, {{0x40010f7638, 0x4}, {0x40010f7688, 0x4}, {0x4001f0b330, 0x1, 0x1}})
	/src/agent/pkg/flow/logging/logger.go:129 +0x328
github.com/grafana/agent/pkg/flow/internal/controller.(*LoggingConfigNode).Evaluate(0x4002cd07e0, 0x4002648000?)
	/src/agent/pkg/flow/internal/controller/node_config_logging.go:67 +0x1e8
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).evaluate(0x4002adc780, {0x8233440, 0x4002cd9d60}, {0x828ea00?, 0x4002cd07e0?})
	/src/agent/pkg/flow/internal/controller/loader.go:789 +0x58
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).Apply.func2({0x823b1c0, 0x4002cd07e0})
	/src/agent/pkg/flow/internal/controller/loader.go:230 +0x7a0
github.com/grafana/agent/pkg/flow/internal/dag.WalkTopological(0x4002cda1a0, {0x4002abaa00, 0x18, 0xc5a3f60?}, 0x400323ef70)
	/src/agent/pkg/flow/internal/dag/walk.go:83 +0x180
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).Apply(0x4002adc780, {0x0, {0x4002ab9200, 0x1a, 0x20}, {0x4002cd6580, 0xf, 0x10}, {0x0, 0x0, ...}, ...})
	/src/agent/pkg/flow/internal/controller/loader.go:181 +0x858
github.com/grafana/agent/pkg/flow.(*Flow).loadSource(0x4002ac81e0, 0x4002a988c0, 0x0, 0x0)
	/src/agent/pkg/flow/flow.go:303 +0x144
github.com/grafana/agent/pkg/flow.(*Flow).LoadSource(...)
	/src/agent/pkg/flow/flow.go:287
github.com/grafana/agent/cmd/internal/flowmode.(*flowRun).Run.func4()
	/src/agent/cmd/internal/flowmode/cmd_run.go:291 +0x1cc
github.com/grafana/agent/cmd/internal/flowmode.(*flowRun).Run(0x4002ac8000, {0xffffd461a1a5, 0x17})
	/src/agent/cmd/internal/flowmode/cmd_run.go:324 +0xcc0
github.com/grafana/agent/cmd/internal/flowmode.runCommand.func1(0x4002ab8300?, {0x4002a846e0?, 0x4?, 0x714a3c4?})
	/src/agent/cmd/internal/flowmode/cmd_run.go:96 +0x38
github.com/spf13/cobra.(*Command).execute(0x4002ac4900, {0x4002a84690, 0x5, 0x5})
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x658
github.com/spf13/cobra.(*Command).ExecuteC(0x4002ac4000)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x320
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/grafana/agent/cmd/internal/flowmode.Run()
	/src/agent/cmd/internal/flowmode/flowmode.go:33 +0x2e8
main.main()
	/src/agent/cmd/grafana-agent/main.go:54 +0x5bc

```

**Comment by rfratto** [comment]
Thanks for the report @christianjedroCDT. Based on the crash stack you provided, the issue doesn't seem to be directly related to `import`, but rather related to the `logging` config block in your main config. Can you share how you have that configured?

**Comment by christianjedroCDT** [comment]
@rfratto i´ve added it after looking at the trace, sorry for that

**Comment by rfratto** [comment]
I have a fix in #6561 that I'm not very pleased with. It'll fix the panic, but given your configuration file I'm not sure how you ended up in a situation where the `logging` block was writing to a null receiver. If there's a deeper problem in the code, it'll likely surface in other ways that the PR won't fix, but hopefully the fix is good enough for now.

**Comment by christianjedroCDT** [comment]
Our configuration file is quite huge and is templated in helm so i tried to condense it as far as possible. This should be the configuration blocks related to logging, maybe this helps more:


```text
remote.kubernetes.secret "credentials" {
  namespace = "opsstack-kubernetes-agent"
  name = "opsstack-kubernetes-agent"
}

remote.kubernetes.configmap "config" {
  namespace = "opsstack-kubernetes-agent"
  name = "opsstack-kubernetes-agent"
}

logging {
    level  = "info"
    format = "json"
    write_to = [loki.write.this.receiver]
}

loki.source.kubernetes_events "this" {
  log_format = "json"
  namespaces = ["opsstack-kubernetes-agent","velero","ingress-nginx","kube-system","default"]
  forward_to = [loki.write.this.receiver]
}

discovery.kubernetes "windows" {
  role = "pod"
  namespaces {
    names         = ["opsstack-kubernetes-agent"]
  }
  selectors {
    role = "pod"
    label = "kubernetes.io/os=windows"
  }
}

loki.source.kubernetes "windows" {
  targets    = discovery.kubernetes.windows.targets
  forward_to = [loki.write.this.receiver]
}

loki.source.api "this" {
  http {
    listen_address = "0.0.0.0"
    listen_port = 9100
  }
  forward_to = [loki.write.this.receiver]
}

loki.write "this" {
  endpoint {
    name = "opsstack"
    url = remote.kubernetes.configmap.config.data["PROMETHEUS_REMOTE_WRITE_URL"] + "/loki/api/v1/push"

    basic_auth {
      username = remote.kubernetes.configmap.config.data["PROMETHEUS_REMOTE_USERNAME"]
      password = remote.kubernetes.secret.credentials.data["PROMETHEUS_REMOTE_PASSWORD"]
    }
  }
}
```

> 3\. I'm noticing that there's an info-level log line despite you setting the logging level to warn. Are you sure you shared the correct config file?

yes thats correct, i used the helm template as reference.

looking at it - could the issue be with 

```
loki.source.kubernetes "windows" {
  targets    = discovery.kubernetes.windows.targets
  forward_to = [loki.write.this.receiver]
}
```

which does not send any data because i´m running in a cluster without windows pods?
